src/Serilog.Extensions.Hosting/Extensions/Hosting/DiagnosticContextCollector.cs

```diff
@@ -0,0 +1,51 @@
+﻿using System;
+using System.Collections.Generic;
+using Serilog.Events;
+
+namespace Serilog.Extensions.Hosting
+{
+    /// <summary>
+    /// A container that receives properties added to a diagnostic context.
+    /// </summary>
+    public sealed class DiagnosticContextCollector : IDisposable
+    {
+        readonly AmbientDiagnosticContextCollector _ambientCollector;
+        List<LogEventProperty> _properties = new List<LogEventProperty>();
```

Comment:
A list, rather than a dictionary, is used to save unnecessary allocations. There's a bit of a leak risk here, and despite the code jumping through a few hoops to mitigate it, we might end up wanting to pay for a dictionary and an extra allocation or two during completion.